File: frontend/src/app/core/api/paged.ts
import { z } from 'zod';

export function paged<T extends z.ZodType>(item: T) {
  return z.object({
    items: z.array(item),
    page: z.number().int().positive(),
    pageSize: z.number().int().positive(),
    total: z.number().int().nonnegative(),
  });
}

export type Paged<T> = {
  items: T[];
  page: number;
  pageSize: number;
  total: number;
};

File: backend/src/Forum.Api/Common/Paged.cs
namespace Forum.Api.Common;

public record Paged<T>(IReadOnlyList<T> Items, int Page, int PageSize, int Total);

public record AuthorSummary(Guid Id, string Username);

public static class Paging
{
    public const int DefaultPageSize = 20;
    public const int MaxPageSize = 100;

    public static (int Page, int PageSize) Clamp(int? page, int? pageSize) =>
        (Math.Max(page ?? 1, 1),
            Math.Clamp(pageSize ?? DefaultPageSize, 1, MaxPageSize));
}
